clippy_lints/src/dereference.rs

```diff
@@ -0,0 +1,71 @@
+use crate::rustc::hir::{Expr, ExprKind, QPath};
+use crate::rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use crate::rustc::{declare_tool_lint, lint_array};
+use if_chain::if_chain;
+use crate::utils::span_lint_and_sugg;
+
+/// **What it does:** Checks for explicit deref() or deref_mut() method calls.
+///
+/// **Why is this bad?** Derefencing by &*x or &mut *x is clearer and more concise,
+/// when not part of a method chain.
+///
+/// **Example:**
+/// ```rust
+/// let b = a.deref();
+/// let c = a.deref_mut();
+///
+/// // excludes
+/// let e = d.deref().unwrap();
```

Comment:
Does your lint excludes these cases? Since `_.deref().len()` gets linted, I have a feeling, that also `_.unwrap()` will get linted.

Reply:
Quite correct - should be the other way round. `d.unwrap().deref()` would not be linted.

Reply:
Could you also add a test for this?

Reply:
https://github.com/rust-lang-nursery/rust-clippy/pull/3258/files#diff-280822c8730d5eac6da4c4b86a9a0fd6R34

https://github.com/rust-lang-nursery/rust-clippy/pull/3258/files#diff-21b45ffb0b35b75693c82b44153ebd60R36

In a method chain, `deref` will only be linted if it is the first method in the chain.